From: Jeff Wang <jeff.wang@airoha.com>
Date: Tue, 26 Mar 2024 13:01:45 +0800
Subject: [PATCH] net: pcs: airoha: Add shared data structure for PCS and PMA

Add a shared data structure to store PCS and PMA regmap pointers for use by
other drivers. This enables better coordination between PCS and common PHY
functions.

--- /dev/null
+++ b/drivers/net/pcs/airoha/common-phy-shared.h
@@ -0,0 +1,11 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (c) 2024 AIROHA Inc
+ * Author: Jeff Wang <jeff.wang@airoha.com>
+ */
+
+#include <linux/regmap.h>
+
+struct airoha_pcs_shared_data {
+	struct regmap *pcs_ana;
+	struct regmap *pcs_pma[2];
+};
--- a/drivers/net/pcs/airoha/pcs-airoha-common.c
+++ b/drivers/net/pcs/airoha/pcs-airoha-common.c
@@ -16,6 +16,7 @@
 #include <linux/reset.h>
 
 #include "pcs-airoha.h"
+#include "common-phy-shared.h"
 
 static void airoha_pcs_setup_scu_eth(struct airoha_pcs_priv *priv,
 				     phy_interface_t interface)
@@ -1030,7 +1031,6 @@ static int airoha_pcs_alloc_maps(struct
 	ret = airoha_pcs_init_named_regmap(pdev, "multi_sgmii", &maps->multi_sgmii);
 	if (ret)
 		return ret;
-
 	ret = airoha_pcs_init_named_regmap(pdev, "usxgmii", &maps->usxgmii_pcs);
 	if (ret)
 		return ret;
@@ -1047,7 +1047,6 @@ static int airoha_pcs_usb_alloc_maps(str
 {
 	struct airoha_pcs_maps *maps = &priv->maps[0];
 	int ret;
-
 	ret = airoha_pcs_init_named_regmap(pdev, "pcs_mac", &maps->pcs_mac);
 	if (ret)
 		return ret;
@@ -1236,6 +1235,7 @@ static int airoha_pcs_probe(struct platf
 	struct fwnode_pcs_provider *provider;
 	struct device *dev = &pdev->dev;
 	struct airoha_pcs_priv *priv;
+	struct airoha_pcs_shared_data *shared;
 	int index, ret;
 
 	data = of_device_get_match_data(dev);
@@ -1292,6 +1292,18 @@ static int airoha_pcs_probe(struct platf
 		break;
 	}
 
+	shared = devm_kzalloc(dev, sizeof(*shared), GFP_KERNEL);
+	if (!shared)
+		return -ENOMEM;
+
+	/* Shared data for common phy */
+	shared->pcs_ana = priv->pcs_ana;
+	memcpy(shared->pcs_pma, priv->pcs_pma, sizeof(shared->pcs_pma));
+	pdev->dev.platform_data = shared;
+	dev_info(dev, "priv->pcs_ana= 0x%lx\n", (unsigned long)priv->pcs_ana);
+	dev_info(dev, "priv->pcs_pma[0]= 0x%lx\n", (unsigned long)priv->pcs_pma[0]);
+	dev_info(dev, "priv->pcs_pma[1]= 0x%lx\n", (unsigned long)priv->pcs_pma[1]);
+
 	if (data->alloc_regmap_fields) {
 		ret = data->alloc_regmap_fields(priv);
 		if (ret)
